### PR TITLE
793: Add a duplicate bulk actions toolbar to the footer.

### LINKED ISSUE #793: Add a second bulk action toolbar at the bottom of the translations list
With the latest GlotDict release the bulk action form is copied also on the footer.
This is very helpful when you are using this workflow: https://github.com/Mte90/GlotDict/issues/103

### PR DESCRIPTION
**Pull request by toolstack**
Resolves #793.

### CODE REVIEW

File: gp-templates/translations.php [hunk 1]
```diff
@@ -25,49 +25,13 @@
...
-	<?php endif; ?>
-	<input type="hidden" name="bulk[redirect_to]" value="<?php echo esc_attr(gp_url_current()); ?>" id="bulk[redirect_to]" />
-	<input type="hidden" name="bulk[row-ids]" value="" id="bulk[row-ids]" />
-	<input type="submit" class="button" value="<?php esc_attr_e( 'Apply', 'glotpress' ); ?>" />
-	</div>
-	<?php gp_route_nonce_field( 'bulk-actions' ); ?>
-</form>
-<?php endif; ?>
-<?php echo gp_pagination( $page, $per_page, $total_translations_count ); ?>
+echo gp_pagination( $page, $per_page, $total_translations_count ); // WPCS: XSS Ok.
```
Comment (ocean90) [inline gp-templates/translations.php]: If `gp_pagination()` is safe we should add to `customAutoEscapedFunctions` in `phpcs.ruleset.xml`

File: gp-templates/helper-functions.php [hunk 2]
```diff
@@ -241,3 +241,56 @@ function references( $project, $entry ) {
...
+	<?php
+	foreach ( GP::$original->get_static( 'priorities' ) as $value => $label ) {
+		$selected = 'normal' === $value ? " selected='selected'" : '';
+		echo "\t<option value='" . esc_attr( $value ) . "' $selected>" . esc_html( $label ) . "</option>\n"; // WPCS: XSS Ok.
+	}
+	?>
+	</select>
+	<?php endif; ?>
+	<input type="hidden" name="bulk[redirect_to]" value="<?php echo esc_attr( gp_url_current() ); ?>" id="bulk[redirect_to]" />
+	<input type="hidden" name="bulk[row-ids]" value="" id="bulk[row-ids]" />
```
Comment (ocean90) [inline gp-templates/helper-functions.php]: This would produce a duplicate ID.

File: gp-templates/helper-functions.php [hunk 3]
```diff
@@ -241,3 +241,56 @@ function references( $project, $entry ) {
...
+	<select name="bulk[priority]" id="bulk-priority-<?php echo $location; // WPCS: XSS Ok. ?>" class="bulk-priority hidden">
+	<?php
+	foreach ( GP::$original->get_static( 'priorities' ) as $value => $label ) {
+		$selected = 'normal' === $value ? " selected='selected'" : '';
+		echo "\t<option value='" . esc_attr( $value ) . "' $selected>" . esc_html( $label ) . "</option>\n"; // WPCS: XSS Ok.
+	}
+	?>
+	</select>
+	<?php endif; ?>
+	<input type="hidden" name="bulk[redirect_to]" value="<?php echo esc_attr( gp_url_current() ); ?>" id="bulk[redirect_to]" />
```
Comment (ocean90) [inline gp-templates/helper-functions.php]: This would produce a duplicate ID.

File: gp-templates/helper-functions.php [hunk 4]
```diff
@@ -241,3 +241,56 @@ function references( $project, $entry ) {
...
+		 */
+		do_action( 'gp_translation_set_bulk_action', $translation_set );
+		?>
+	</select>
+	<?php if ( $can_write ) : ?>
+	<select name="bulk[priority]" id="bulk-priority-<?php echo $location; // WPCS: XSS Ok. ?>" class="bulk-priority hidden">
+	<?php
+	foreach ( GP::$original->get_static( 'priorities' ) as $value => $label ) {
+		$selected = 'normal' === $value ? " selected='selected'" : '';
+		echo "\t<option value='" . esc_attr( $value ) . "' $selected>" . esc_html( $label ) . "</option>\n"; // WPCS: XSS Ok.
```
Comment (ocean90) [inline gp-templates/helper-functions.php]: `$selected` can be replaced with `selected()`.

File: gp-templates/helper-functions.php [hunk 5]
```diff
@@ -241,3 +241,56 @@ function references( $project, $entry ) {
...
+/**
+ * Output the bulk actions toolbar in the translations page.
+ *
+ * @param string $bulk_action  The URL to submit the form to.
+ * @param string $can_write    Can the current user write translations to the database.
+ * @param string $location     The location of this toolbar, used to make id's unique for each instance on a page.
+ */
+function gp_translations_bulk_actions_toolbar( $bulk_action, $can_write, $location = 'top' ) {
+?>
+<form id="bulk-actions-toolbar-<?php echo $location; // WPCS: XSS Ok. ?>" class="filters-toolbar bulk-actions" action="<?php echo $bulk_action; // WPCS: XSS Ok. ?>" method="post">
```
Comment (ocean90) [inline gp-templates/helper-functions.php]: `XSS Ok` 👎, can be easily fixed by using `esc_attr()`.

**Review by ocean90 — CHANGES_REQUESTED** [review]

File: gp-templates/translations.php (hunk 1 above)
Comment (toolstack) [reply gp-templates/translations.php]: As safe as anything that has a filter in it ;)

I'll add it to the ruleset.

File: gp-templates/helper-functions.php (hunk 5 above)
Comment (toolstack) [reply gp-templates/helper-functions.php]: Converted to `esc_attr()`.

File: gp-templates/helper-functions.php (hunk 3 above)
Comment (toolstack) [reply gp-templates/helper-functions.php]: Updated ID to include the `$location` var as the ID is not used anywhere in GP.

File: gp-templates/helper-functions.php (hunk 2 above)
Comment (toolstack) [reply gp-templates/helper-functions.php]: Updated ID to include the `$location` var as the ID is not used anywhere in GP.

File: gp-templates/helper-functions.php (hunk 4 above)
Comment (toolstack) [reply gp-templates/helper-functions.php]: Updated.

File: gp-templates/helper-functions.php [hunk 6]
```diff
@@ -241,3 +241,55 @@ function references( $project, $entry ) {
...
+		echo "\t<option value='" . esc_attr( $value ) . "' " . selected( 'normal', $value, false )  . ">" . esc_html( $label ) . "</option>\n"; // WPCS: XSS Ok.
+	}
+	?>
+	</select>
+	<?php endif; ?>
+	<input type="hidden" name="bulk[redirect_to]" value="<?php echo esc_attr( gp_url_current() ); ?>" id="bulk-redirect_to-<?php echo esc_attr( $location ); ?>" />
+	<input type="hidden" name="bulk[row-ids]" value="" id="bulk-row-ids-<?php echo esc_attr( $location ); ?>" />
+	<input type="submit" class="button" value="<?php esc_attr_e( 'Apply', 'glotpress' ); ?>" />
+	</div>
+	<?php gp_route_nonce_field( 'bulk-actions' ); ?>
```
Comment (ocean90) [inline gp-templates/helper-functions.php]: This one uses `wp_nonce_field()` which also adds an `id` attribute to the input field. Not sure how to fix that though.

**Review by ocean90 — CHANGES_REQUESTED** [review]

File: gp-templates/helper-functions.php (hunk 6 above)
Comment (toolstack) [reply gp-templates/helper-functions.php]: That's relatively easy to fix, we can fudge the ID before echoing it back out, I'll commit a fix.

**Comment by toolstack** [comment]
@ocean90 I've tracked down a few of the errors (PHPCS was throwing the wrong line numbers for some reason) but for the life of me I can't find the last one.

Unless you have some thoughts on it I'm going to suggest we merge it as is.

File: gp-templates/helper-functions.php [hunk 7]
```diff
@@ -258,3 +258,68 @@ function references( $project, $entry ) {
...
+		 * Fires inside the bulk action menu for translation sets.
+		 *
+		 * Printing out option elements here will add those to the translation
+		 * bulk options drop down menu.
+		 *
+		 * @since 1.0.0
+		 *
+		 * @param GP_Translation_Set $set The translation set.
+		 */
+		do_action( 'gp_translation_set_bulk_action', $translation_set );
```
Comment (ocean90) [inline gp-templates/helper-functions.php:294]: There is no such variable.

File: gp-templates/helper-functions.php [hunk 8]
```diff
@@ -258,3 +258,68 @@ function references( $project, $entry ) {
...
+	</select>
+	<?php if ( $can_write ) : ?>
+	<select name="bulk[priority]" id="bulk-priority-<?php echo esc_attr( $location ); ?>" class="bulk-priority hidden">
+	<?php
+	$labels = [
+		'hidden' => _x( 'hidden', 'Priority', 'glotpress' ),
+		'low'    => _x( 'low', 'Priority', 'glotpress' ),
+		'normal' => _x( 'normal', 'Priority', 'glotpress' ),
+		'high'   => _x( 'high', 'Priority', 'glotpress' ),
+	];
```
Comment (ocean90) [inline gp-templates/helper-functions.php:305]: This is unused.

File: gp-templates/helper-functions.php [hunk 9]
```diff
@@ -258,3 +258,68 @@ function references( $project, $entry ) {
...
+	<input type="hidden" name="bulk[redirect_to]" value="<?php echo esc_attr( gp_url_current() ); ?>" id="bulk-redirect_to-<?php echo esc_attr( $location ); ?>" />
+	<input type="hidden" name="bulk[row-ids]" value="" id="bulk-row-ids-<?php echo esc_attr( $location ); ?>" />
+	<input type="submit" class="button" value="<?php esc_attr_e( 'Apply', 'glotpress' ); ?>" />
+	</div>
+	<?php
+		$nonce = gp_route_nonce_field( 'bulk-actions', false );
+
+		$nonce = str_replace( 'id="_gp_route_nonce"', 'id="_gp_route_nonce_' . esc_attr( $location ) . '"', $nonce );
+
+		echo $nonce; // WPCS: XSS Ok.
```
Comment (ocean90) [inline gp-templates/helper-functions.php]: The empty lines should be removed.

File: gp-templates/translations.php [hunk 10]
```diff
@@ -290,7 +245,11 @@
 	endif;
 ?>
 </table>
-<?php echo gp_pagination( $page, $per_page, $total_translations_count ); ?>
+<?php
+gp_translations_bulk_actions_toolbar( $bulk_action, $can_write, 'bottom' );
+
+echo gp_pagination( $page, $per_page, $total_translations_count ); // WPCS: XSS Ok.
```
Comment (ocean90) [inline gp-templates/translations.php]: `// WPCS: XSS Ok.` shouldn't be necessary anymore since it's now part of `customAutoEscapedFunctions`.

**Review by ocean90 — CHANGES_REQUESTED** [review]

File: gp-templates/helper-functions.php (hunk 7 above)
Comment (toolstack) [reply gp-templates/helper-functions.php:294]: Added it to the function call.

File: gp-templates/helper-functions.php (hunk 8 above)
Comment (toolstack) [reply gp-templates/helper-functions.php:305]: It is in the loop below this line.

File: gp-templates/helper-functions.php (hunk 9 above)
Comment (toolstack) [reply gp-templates/helper-functions.php]: Removed.

File: gp-templates/translations.php (hunk 10 above)
Comment (toolstack) [reply gp-templates/translations.php]: Removed.

File: gp-templates/helper-functions.php (hunk 8 above)
Comment (toolstack) [reply gp-templates/helper-functions.php:305]: Fixed.

**Comment by toolstack** [comment]
@ocean90 looks like this is good to go now.

**Review by ocean90 — APPROVED** [review]

**Pull request merged**